Node e2e memory eviction test

mtaufen opened the pull request:
This tests memory evictions.
See related issue #28619 and fix to cadvisor https://github.com/google/cadvisor/pull/1380.

cc @vishh @derekwaynecarr @timstclair

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28693)

<!-- Reviewable:end -->

mtaufen committed:
Node e2e memory eviction test

This test creates three pods with QoS of besteffort, burstable, and
guaranteed, respectively, which each contain a container that tries to
consume almost all the available memory at a rate of about 12Mi/10sec.

The expectation is that eviction will be initiated when the hard
memory.available<250Mi threshold is triggered, and that eviction will proceed
in the order of besteffort, then burstable. Since guaranteed pods should
only be evicted if something charged to the host uses more resources
than were reserved for it, we currently end the test when besteffort and
burstable have both been evicted.

Note that this commit also sets --eviction-hard=memory.available<250Mi
to enable eviction during tests.

vishh commented:
@k8s-bot node e2e test this github issue #IGNORE

pwittrock commented:
@vishh Reassigning to you since I have a PR backlog

mtaufen commented:
@vishh PTAL

derekwaynecarr commented on test/e2e_node/memory_eviction_test.go:
```diff
@@ -0,0 +1,191 @@
+
+})
+
+func createMemhogPod(f *framework.Framework, genName string, ctnName string, res api.ResourceRequirements) *api.Pod {
```
so I have an image that uses `memhog` itself (derekwaynecarr/memhog).

you can see some sample pods using it here: 
https://github.com/derekwaynecarr/kubernetes/tree/examples-eviction/demo/kubelet-eviction

maybe we can just have a dedicated image we share?

vishh replied:
On that note, I created vish/stress because memhog was a large image and it does not let us control the rate of consumption of memory. 

$ docker run --rm vish/stress --help
Usage of /stress:
  -alsologtostderr
    	log to standard error as well as files
  -cpus int
    	total number of CPUs to utilize
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace (default :0)
  -log_dir string
    	If non-empty, write log files in this directory
  -logtostderr
    	log to standard error instead of files
  -mem-alloc-size string
    	amount of memory to be consumed in each allocation (default "4Ki")
  -mem-alloc-sleep duration
    	duration to sleep between allocations (default 1ms)
  -mem-total string
    	total memory to be consumed. Memory will be consumed via multiple allocations.
  -stderrthreshold value
    	logs at or above this threshold go to stderr
  -v value
    	log level for V logs
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging

vishh commented:
@mtaufen If you can address the pod image comments, this PR can go in.

mtaufen commented:
Rebased and transitioned to @vishh's image. But I just saw a burstable fail before a best-effort while running the test. I used to set `--eviction-hard` to `memory.available<500Mi` in `e2e_service.go`, but now it's set via a flag for the test context (see `test/e2e/framework/test_context.go.RegisterNodeFlags`), and doesn't appear to be set by default. I think not setting it by default probably results in eviction remaining disabled as a feature for e2e_node tests. I just changed the default to what I used to set in `e2e_service.go` and am re-testing to see if that helps.

mtaufen commented:
I've run a few scenarios to start getting a ballpark for the thresholds where I see proper eviction order and where I don't (I'm operating under the assumption that when I don't, it's due to an OOM-kill, but I haven't actually investigated that yet):

```
Passed:
--eviction-hard=memory.available<500Mi x1
--eviction-hard=memory.available<300Mi x1
--eviction-hard=memory.available<250Mi x1
--eviction-hard=memory.available<230Mi x1

Failed:
--eviction-hard=memory.available<200Mi x2
--eviction-hard=memory.available<100Mi x1
```

It seems that failures start to crop up somewhere between a threshold of `200Mi` and `230Mi`, currently testing `220Mi` to see what happens there.

vishh commented:
I'm ok with the thresholds being higher for now. We can lower them in a
subsequent PR.

On Mon, Aug 8, 2016 at 8:36 AM, Michael Taufen <notifications@github.com>
wrote:

> I've run a few scenarios to start getting a ballpark for the thresholds
> where I see proper eviction order and where I don't (I'm operating under
> the assumption that when I don't, it's due to an OOM-kill, but I haven't
> actually investigated that yet):
>
> Passed:
> --eviction-hard=memory.available<500Mi x1
> --eviction-hard=memory.available<300Mi x1
> --eviction-hard=memory.available<250Mi x1
> --eviction-hard=memory.available<230Mi x1
>
> Failed:
> --eviction-hard=memory.available<200Mi x2
> --eviction-hard=memory.available<100Mi x1
>
> It seems that failures start to crop up somewhere between a threshold of
> 200Mi and 230Mi, currently testing 220Mi to see what happens there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28693#issuecomment-238275862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKIOZkKQBYOoEZun5P4Jr42tSkeHFks5qd0z3gaJpZM4JIMYb>
> .
>

mtaufen commented:
Ok, set to `250Mi` for now.

mtaufen commented:
@vishh PTAL this can probably go in.
